File: modules/desktop/src/interpreter/stepA_mal.js
if (typeof module !== 'undefined') {
    var types = require('./types');
    // var readline = require('./node_readline');
    var readline = require('readline');
    var reader = require('./reader');
    var printer = require('./printer');
    var Env = require('./env').Env;
    var core = require('./core');
    var EventEmitter = require('events');   var net = require('net')
}

// read
function READ(str) {
    return reader.read_str(str);
}

// eval
function qqLoop (acc, elt) {
    if (types._list_Q(elt) && elt.length
        && types._symbol_Q(elt[0]) && elt[0].value == 'splice-unquote') {
        return [types._symbol("concat"), elt[1], acc];
    } else {
        return [types._symbol("cons"), quasiquote (elt), acc];
    }
}
function quasiquote(ast) {
    if (types._list_Q(ast) && 0<ast.length
        && types._symbol_Q(ast[0]) && ast[0].value == 'unquote') {
        return ast[1];
    } else if (types._list_Q(ast)) {
        return ast.reduceRight(qqLoop,[]);
    } else if (types._vector_Q(ast)) {
        return [types._symbol("vec"), ast.reduceRight(qqLoop,[])];
    } else if (types._symbol_Q(ast) || types._hash_map_Q(ast)) {
        return [types._symbol("quote"), ast];
    } else {
        return ast;
    }
}

function is_macro_call(ast, env) {
    return types._list_Q(ast) &&
           types._symbol_Q(ast[0]) &&
           env.find(ast[0]) &&
           env.get(ast[0])._ismacro_;
}

function macroexpand(ast, env) {
    while (is_macro_call(ast, env)) {
        var mac = env.get(ast[0]);
        ast = mac.apply(mac, ast.slice(1));
    }
    return ast;
}

function eval_ast(ast, env) {
    if (types._symbol_Q(ast)) {
        return env.get(ast);
    } else if (types._list_Q(ast)) {
        return ast.map(function(a) { return EVAL(a, env); });
    } else if (types._vector_Q(ast)) {
        var v = ast.map(function(a) { return EVAL(a, env); });
        v.__isvector__ = true;
        return v;
    } else if (types._hash_map_Q(ast)) {
        var new_hm = {};
        for (k in ast) {
            new_hm[k] = EVAL(ast[k], env);
        }
        return new_hm;
    } else {
        return ast;
    }
}

function _EVAL(ast, env) {
    while (true) {

    //printer.println("EVAL:", printer._pr_str(ast, true));
    if (!types._list_Q(ast)) {
        return eval_ast(ast, env);
    }

    // apply list
    ast = macroexpand(ast, env);
    if (!types._list_Q(ast)) {
        return eval_ast(ast, env);
    }
    if (ast.length === 0) {
        return ast;
    }

    var a0 = ast[0], a1 = ast[1], a2 = ast[2], a3 = ast[3];
    switch (a0.value) {
    case "def!":
        var res = EVAL(a2, env);
        return env.set(a1, res);
    case "let*":
        var let_env = new Env(env);
        for (var i=0; i < a1.length; i+=2) {
            let_env.set(a1[i], EVAL(a1[i+1], let_env));
        }
        ast = a2;
        env = let_env;
        break;
    case "quote":
        return a1;
    case "quasiquoteexpand":
        return quasiquote(a1);
    case "quasiquote":
        ast = quasiquote(a1);
        break;
    case 'defmacro!':
        var func = types._clone(EVAL(a2, env));
        func._ismacro_ = true;
        return env.set(a1, func);
    case 'macroexpand':
        return macroexpand(a1, env);
    case "try*":
        try {
            return EVAL(a1, env);
        } catch (exc) {
            if (a2 && a2[0].value === "catch*") {
                if (exc instanceof Error) { exc = exc.message; }
                return EVAL(a2[2], new Env(env, [a2[1]], [exc]));
            } else {
                throw exc;
            }
        }
    case "do":
        eval_ast(ast.slice(1, -1), env);
        ast = ast[ast.length-1];
        break;
    case "if":
        var cond = EVAL(a1, env);
        if (cond === null || cond === false) {
            ast = (typeof a3 !== "undefined") ? a3 : null;
        } else {
            ast = a2;
        }
        break;
    case "fn*":
        return types._function(EVAL, Env, a2, env, a1);
    default:
        var el = eval_ast(ast, env), f = el[0];
        if (f.__ast__) {
            ast = f.__ast__;
            env = f.__gen_env__(el.slice(1));
        } else {
            return f.apply(f, el.slice(1));
        }
    }

    }
}

function EVAL(ast, env) {
    var result = _EVAL(ast, env);
    return (typeof result !== "undefined") ? result : null;
}

// print
function PRINT(exp) {
    return printer._pr_str(exp, true);
}

// repl
var repl_env = new Env();
var rep = function(str) { return PRINT(EVAL(READ(str), repl_env)); };

// core.js: defined using javascript
for (var n in core.ns) { repl_env.set(types._symbol(n), core.ns[n]); }
repl_env.set(types._symbol('eval'), function(ast) {
    return EVAL(ast, repl_env); });
repl_env.set(types._symbol('*ARGV*'), []);

// core.mal: defined using the language itself
rep("(def! *host-language* \"javascript\")")
rep("(def! not (fn* (a) (if a false true)))");
rep("(def! load-file (fn* (f) (eval (read-string (str \"(do \" (slurp f) \"\nnil)\")))))");
rep("(defmacro! cond (fn* (& xs) (if (> (count xs) 0) (list 'if (first xs) (if (> (count xs) 1) (nth xs 1) (throw \"odd number of forms to cond\")) (cons 'cond (rest (rest xs)))))))");


const emitter1 = new EventEmitter();
function send_msg(env, exp="") {
    sendMessage(env, exp);
    emitter1.once('data', data => {
        console.log(data);
    })
    // return new Promise(resolve => {
    // })
}

function sendMessage(env, exp = "") {
    exp = printer._pr_str(exp)
    let client_socket = new net.Socket();
    client_socket.connect(env[1], env[0]);
    client_socket.write(exp.toString());

    client_socket.on('data', (msg) => {
        emitter1.emit('data', msg.toString())
    });
}

export function run_command(line) {
    try {
        if (line) { return rep(line); }
    } catch (exc) {
        if (exc instanceof reader.BlankException) {
            return ""; 
            }
        if (exc instanceof Error) { 
            return exc.stack
        }
        else { 
            return "Error: " + exc
        }
    }
}

function server(host, port=8124) {
    const socket = net.createServer((socket) => {
        console.log(`(${socket.remoteAddress}:${socket.remotePort})`);
        const emitter = new EventEmitter();

        emitter.on('data', (data) => {
            console.log("Exp: " + data.toString());
            let response = run_command(data.toString())
            console.log("Result: " + response);
            socket.write(response);
        });

        socket.on('data', (data) => {
            emitter.emit('data', data);
        });
    });

    socket.on('error', (err) => {
        console.log(err);
    });

    socket.listen(port, host, () => {
        console.log('Server run at: ' + port);
    });
}

function repl(prompt=">>> ") {
    if (typeof process !== 'undefined' && process.argv.length > 2) {
        repl_env.set(types._symbol('*ARGV*'), process.argv.slice(3));
        rep('(load-file "' + process.argv[2] + '")');
        process.exit(0);
    }

    rep("(println (str \"Mal [\" *host-language* \"]\"))");
    const rl = readline.createInterface({
        input: process.stdin,
        output: process.stdout,
    });
    rl.setPrompt(prompt);
    rl.prompt()
    rl.on('line',(line) => {
        if (line === null) { 
            rl.close();
        }
        try {
            if (line) { printer.println(rep(line)); }
        } catch (exc) {
            if (exc instanceof Error) { console.warn(exc.stack) }
            else { console.warn("Error: " + printer._pr_str(exc, true)) }
        }
        rl.prompt()
    })
}

function load_lib(file_path) {
    let mod = require(file_path.replace(/\.[^/.]+$/, ""));
    Object.entries(mod).forEach(([key, value]) => {
        if (typeof value === 'function') {
            console.log(key)
            repl_env.set(types._symbol(key.replace(/_/g, "-")), value)
        }
    })
}

repl_env.set(types._symbol('send-msg'),send_msg)
repl_env.set(types._symbol('server'), server)
repl_env.set(types._symbol('repl'), repl)
repl_env.set(types._symbol('global-symbols-string'), ()=> {return Object.entries(repl_env.data).map(([key, value]) => key)})
repl_env.set(types._symbol('exit'), process.exit)
repl_env.set(types._symbol('load-lib'), load_lib)

// export {
//     repl_env,
//     server,
//     repl,
//     rep,
//     run_command
// }

// repl()

File: python/interpreter/stepA_mal.py
import functools
import sys, traceback
import mal_readline
import mal_types as types
import reader, printer
from env import Env
import core
import socket
import importlib
import os


# read
def READ(str):
    return reader.read_str(str)

# eval
def qq_loop(acc, elt):
    if types._list_Q(elt) and len(elt) == 2 and elt[0] == u'splice-unquote':
        return types._list(types._symbol(u'concat'), elt[1], acc)
    else:
        return types._list(types._symbol(u'cons'), quasiquote(elt), acc)

def qq_foldr(seq):
    return functools.reduce(qq_loop, reversed(seq), types._list())

def quasiquote(ast):
    if types._list_Q(ast):
        if len(ast) == 2 and ast[0] == u'unquote':
            return ast[1]
        else:
            return qq_foldr(ast)
    elif types._hash_map_Q(ast) or types._symbol_Q(ast):
        return types._list(types._symbol(u'quote'), ast)
    elif types._vector_Q (ast):
        return types._list(types._symbol(u'vec'), qq_foldr(ast))
    else:
        return ast

def is_macro_call(ast, env):
    return (types._list_Q(ast) and
            types._symbol_Q(ast[0]) and
            env.find(ast[0]) and
            hasattr(env.get(ast[0]), '_ismacro_'))

def macroexpand(ast, env):
    while is_macro_call(ast, env):
        mac = env.get(ast[0])
        ast = mac(*ast[1:])
    return ast

def eval_ast(ast, env):
    if types._symbol_Q(ast):
        return env.get(ast)
    elif types._list_Q(ast):
        return types._list(*map(lambda a: EVAL(a, env), ast))
    elif types._vector_Q(ast):
        return types._vector(*map(lambda a: EVAL(a, env), ast))
    elif types._hash_map_Q(ast):
        return types.Hash_Map((k, EVAL(v, env)) for k, v in ast.items())
    else:
        return ast  # primitive value, return unchanged

def EVAL(ast, env):
    while True:
        #print("EVAL %s" % printer._pr_str(ast))
        if not types._list_Q(ast):
            return eval_ast(ast, env)

        # apply list
        ast = macroexpand(ast, env)
        if not types._list_Q(ast):
            return eval_ast(ast, env)
        if len(ast) == 0: return ast
        a0 = ast[0]

        if "def!" == a0:
            a1, a2 = ast[1], ast[2]
            res = EVAL(a2, env)
            return env.set(a1, res)
        elif "let*" == a0:
            a1, a2 = ast[1], ast[2]
            let_env = Env(env)
            for i in range(0, len(a1), 2):
                let_env.set(a1[i], EVAL(a1[i+1], let_env))
            ast = a2
            env = let_env
            # Continue loop (TCO)
        elif "quote" == a0:
            return ast[1]
        elif "quasiquoteexpand" == a0:
            return quasiquote(ast[1]);
        elif "quasiquote" == a0:
            ast = quasiquote(ast[1]);
            # Continue loop (TCO)
        elif 'defmacro!' == a0:
            func = types._clone(EVAL(ast[2], env))
            func._ismacro_ = True
            return env.set(ast[1], func)
        elif 'macroexpand' == a0:
            return macroexpand(ast[1], env)
        elif "py!*" == a0:
            exec(compile(ast[1], '', 'single'), globals())
            return None
        elif "py*" == a0:
            return types.py_to_mal(eval(ast[1]))
        elif "." == a0:
            el = eval_ast(ast[2:], env)
            f = eval(ast[1])
            return f(*el)
        elif "try*" == a0:
            if len(ast) < 3:
                return EVAL(ast[1], env)
            a1, a2 = ast[1], ast[2]
            if a2[0] == "catch*":
                err = None
                try:
                    return EVAL(a1, env)
                except types.MalException as exc:
                    err = exc.object
                except Exception as exc:
                    err = exc.args[0]
                catch_env = Env(env, [a2[1]], [err])
                return EVAL(a2[2], catch_env)
            else:
                return EVAL(a1, env);
        elif "do" == a0:
            eval_ast(ast[1:-1], env)
            ast = ast[-1]
            # Continue loop (TCO)
        elif "if" == a0:
            a1, a2 = ast[1], ast[2]
            cond = EVAL(a1, env)
            if cond is None or cond is False:
                if len(ast) > 3: ast = ast[3]
                else:            ast = None
            else:
                ast = a2
            # Continue loop (TCO)
        elif "fn*" == a0:
            a1, a2 = ast[1], ast[2]
            return types._function(EVAL, Env, a2, env, a1)
        else:
            el = eval_ast(ast, env)
            f = el[0]
            if hasattr(f, '__ast__'):
                ast = f.__ast__
                env = f.__gen_env__(el[1:])
            else:
                return f(*el[1:])

# print
def PRINT(exp):
    return printer._pr_str(exp)

# repl
repl_env = Env()
def REP(str):
    return PRINT(EVAL(READ(str), repl_env))

# core.py: defined using python
for k, v in core.ns.items(): repl_env.set(types._symbol(k), v)
repl_env.set(types._symbol('eval'), lambda ast: EVAL(ast, repl_env))
repl_env.set(types._symbol('*ARGV*'), types._list(*sys.argv[2:]))

# core.mal: defined using the language itself
REP("(def! *host-language* \"python\")")
REP("(def! not (fn* (a) (if a false true)))")
REP("(def! load-file (fn* (f) (eval (read-string (str \"(do \" (slurp f) \"\nnil)\")))))")
REP("(defmacro! cond (fn* (& xs) (if (> (count xs) 0) (list 'if (first xs) (if (> (count xs) 1) (nth xs 1) (throw \"odd number of forms to cond\")) (cons 'cond (rest (rest xs)))))))")


def send_msg(env, msg, wait=True):
    exp = printer._pr_str(msg)
    client_socket = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
    client_socket.connect((env[0], env[1]))
    client_socket.sendall(exp.encode('utf-8'))
    if wait:
        data = b'' 
        while True:
            msg = client_socket.recv(1024)
            data += msg
            if len(data) < 1024:
                client_socket.close()
                break
        return data.decode('utf-8')
    else: 
        return ""

def run_command(line):
    try:
        return REP(line)
    except reader.Blank: return "" 
    except types.MalException as e:
        # return "Error:", printer._pr_str(e.object)
        return "Error: "+ e.object
    except Exception as e:
        return str("Exception: "+ str(e))

def server(host, port):
    server_socket = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
    server_socket.bind((host, port))
    server_socket.listen(5)
    # print("Server run at: " + str(host) + ":" + str(port))

    while True:
        client_sockt, addr = server_socket.accept()
        exp = b''
        while True:
            data = client_sockt.recv(1024)
            exp += data
            if len(data) < 1024:
                break
        exp = exp.decode('utf-8')
        result = run_command(exp)
        client_sockt.sendall(result.encode('utf-8'))
        # print(str(addr))
        # print("Exp: " + str(exp))
        # print("Result: " + result)
        client_sockt.close()

def repl(prompt='>>> '):
    if len(sys.argv) >= 2:
        REP('(load-file "' + sys.argv[1] + '")')
        sys.exit(0)

    # repl loop
    REP("(println (str \"Mal [\" *host-language* \"]\"))")
    while True:
            line = mal_readline.readline(prompt)
            result = run_command(line)
            print(result)

def load_lib(file_path, package=None):
    if package:
        file_path ="."+ os.path.relpath(file_path, os.path.dirname(__file__))
        module_name = file_path.replace('\\', '.').replace('.py', '')
        module = importlib.import_module(package+module_name)
    else:
        module_name = file_path.replace('\\', '.').replace('.py', '')
        module = importlib.import_module(module_name)
    local = vars(module)
    local = {k.replace("_", "-") if not (k.startswith("__") and k.endswith("__")) else k: v for k, v in local.items()}
    for key, value in local.items():
        repl_env.set(types._symbol(key), value)

repl_env.set(types._symbol('send-msg'), send_msg)
repl_env.set(types._symbol('server'), server)
repl_env.set(types._symbol('repl'), repl)
repl_env.set(types._symbol('global-symbols-string'), lambda : list(repl_env.data.keys()))
repl_env.set(types._symbol('exit'), sys.exit)
repl_env.set(types._symbol('load-lib'), load_lib)

# repl()